File: Java/src/main/java/org/apd/executor/ReadPreferredExecutor.java
package org.apd.executor;

import org.apd.storage.SharedDatabase;

import java.util.concurrent.Semaphore;

public class ReadPreferredExecutor {
    private static Semaphore resourceLock = new Semaphore(1);
    private static Semaphore readersSemaphore = new Semaphore(1);
    private static int readerCount = 0;
    private static SharedDatabase sharedDatabase;

    public ReadPreferredExecutor(SharedDatabase sharedDatabase) {
        this.sharedDatabase = sharedDatabase;
    }

    public void read(int index) {
        try {
            readersSemaphore.acquire();

            if (++readerCount == 1) {
                resourceLock.acquire();
            }

            readersSemaphore.release();

            sharedDatabase.getData(index);

            readersSemaphore.acquire();

            if (--readerCount == 0) {
                resourceLock.release();
            }

            readersSemaphore.release();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public void write(int index, String data) {
        try {
            resourceLock.acquire();

            sharedDatabase.addData(index, data);

            resourceLock.release();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: C#/MultiThread/ReadPreferredExecutor.cs
using System;
using System.Diagnostics;
using MultiThread.Data;
using System.Threading;

namespace MultiThread
{
    public class ReadPreferredExecutor
    {
        private readonly SharedDatabase _sharedDatabase;

        private static readonly SemaphoreSlim resourceLock = new SemaphoreSlim(1, 1);
        private static readonly SemaphoreSlim readersSemaphore = new SemaphoreSlim(1, 1);
        private static int readerCount = 0;

        public ReadPreferredExecutor(SharedDatabase sharedDatabase)
        {
            _sharedDatabase = sharedDatabase;
        }

        public void Read(int index)
        {
            try
            {
                readersSemaphore.Wait();

                if (++readerCount == 1)
                {
                    resourceLock.Wait();
                }

                readersSemaphore.Release();

                _sharedDatabase.GetData(index);

                readersSemaphore.Wait();
                if (--readerCount == 0)
                {
                    resourceLock.Release();
                }

                readersSemaphore.Release();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        public void Write(int index, string data)
        {
            try
            {
                resourceLock.Wait();

                _sharedDatabase.AddData(index, data);

                resourceLock.Release();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }
    }
}